### PR TITLE
Implemented safer C++ DataContainer data type handling, fixes #1193

### LINKED ISSUE #1193: Longer term solutions to DataContainer data type issue needed.
Some `DataContainer` methods compute a value that is complex for Gadgetron containers and real for other containers, and so, the variable that would store such a value cannot be explicitly declared in this abstract base class. Currently, this issue is quick-fixed by passing a `void*` to these methods that is cast to `float*` or `complex_float_t*` in the respective methods of the derived classes.

Possible longer term fixes:
- use the largest type in the C++ interface (i.e. std::complex<double>) as opposed to `void *`. Easy, but probably unexpected.
- use `std::any` (and if needed with a virtual function `std::any DataContainer::get_zero_data_element()` or so). More complicated and not very elegant.
- template `DataContainer` if possible.

### PR DESCRIPTION
**Pull request by evgueni-ovtchinnikov**
## Changes in this pull request

Current C++ `DataContainer` classes use static casts of `void*` pointers, which may be unsafe (cf. #1193).

This PR implements an alternative template-based design whereby static casts are only used on C interface level, which is not supposed to be used by SIRF users. 

## Related issues

Fixes #1193

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] I have added docstrings/doxygen in line with the guidance in the developer guide
- [x] I have implemented unit tests that cover any new or modified functionality
- [x] The code builds and runs on my machine
- [ ] CHANGES.md has been updated with any functionality change

## Contribution Notes

Please read and adhere to the [contribution guidelines](https://github.com/SyneRBI/SIRF/blob/master/CONTRIBUTING.md).

Please tick the following: 

 - [x] The content of this Pull Request (the Contribution) is intentionally submitted for inclusion in SIRF (the Work) under the terms and conditions of the [Apache-2.0](https://www.apache.org/licenses/LICENSE-2.0) License.

### CODE REVIEW

File: src/Registration/cReg/NiftyResampler.cpp [hunk 1]
```diff
@@ -240,7 +240,7 @@ static void set_post_resample_outputs(std::shared_ptr<DataContainer> &out_to_ret
         NumberType::Type output_num_type = (*output_to_return_sptr->begin()).get_typeID();
         if (output_num_type != NumberType::CXFLOAT)
             throw std::runtime_error("NiftyResampler: Only complex type currently supported is complex float");
-        ImageData<dataType> ::Iterator &it_out = output_to_return_sptr->begin();
+        typename ImageData<dataType> ::Iterator &it_out = output_to_return_sptr->begin();
```
Comment (KrisThielemans) [inline src/Registration/cReg/NiftyResampler.cpp:243]: much simpler to say
```suggestion
        auto &it_out = output_to_return_sptr->begin();
```
but are you sure it needs to be a reference? It's going to change, and therefore affect `output_to_return_sptr->begin()`, which seems terribly dangerous. (This is actually also true for `it_real` etc)

Comment (evgueni-ovtchinnikov) [reply src/Registration/cReg/NiftyResampler.cpp:243]: I cannot be sure about anything here - this is Richard's code, I simply had to add `typename` for SIRF to build (now it builds but synergistic tests fail).

Is `auto` clever enough to use `typename`?

File: src/common/include/sirf/common/DataContainer.h [hunk 2]
```diff
@@ -37,11 +37,239 @@ which rely on the same features of the items.
...
+
+		/// \c *this = elementwise sum of elementwise \c x*a and \c y*b
+		void xapyb(
+			const DataContainer& a_x, const DataContainer& a_a,
+			const DataContainer& a_y, const void* ptr_b)
+		{
+			xapyb(a_y, ptr_b, a_x, a_a);
+		}
+
+		static T product(T x, T y)
```
Comment (KrisThielemans) [inline src/common/include/sirf/common/DataContainer.h:196]: I think that all these `static` function have to be `protected` (or even `private`). they're not part of the public interface really

File: src/xGadgetron/cGadgetron/gadgetron_data_containers.cpp [hunk 3]
```diff
@@ -1571,31 +1577,31 @@ GadgetronImageData::multiply(const DataContainer& a_x, const DataContainer& a_y)
 {
 	SIRF_DYNAMIC_CAST(const GadgetronImageData, x, a_x);
 	SIRF_DYNAMIC_CAST(const GadgetronImageData, y, a_y);
-    binary_op(x, y, DataContainer::product<complex_float_t>);
+    binary_op(x, y, DataContainerTempl<complex_float_t>::product);
```
Comment (KrisThielemans) [inline src/xGadgetron/cGadgetron/gadgetron_data_containers.cpp:1580]: I think all these would be a lot cleaner if we'd have
```
typedef GadgetronImageData self_type;
```
as then we can do
```
    SIRF_DYNAMIC_CAST(const  self_type, x, a_x);
    SIRF_DYNAMIC_CAST(const self_type, y, a_y);
    binary_op(x, y, self_type::product);
```
such that the code looks class-independent

**Review by KrisThielemans — COMMENTED** [review]
some probably mostly random comments but hopefully helpful in this review.

I'm not sure if templating `ImageData` is a good idea. It certainly breaks backwards compatibility in a huge way. Possibly this could be done by having an intermediate `ImageDataTemplate` like you did for DataContainer`. However, it might not work, as I guess we'd then need to derive `ImageData` from `DataContainer`, but `ImageDataTemplate` from `DataContainerTemplate`. This can be done in C++, but has the potential for some confusion.

File: src/Registration/cReg/NiftyResampler.cpp (hunk 1 above)
Comment (evgueni-ovtchinnikov) [reply src/Registration/cReg/NiftyResampler.cpp]: virtual method's return types must be covariant, must they not?

File: src/Registration/cReg/cReg_p.cpp [hunk 4]
```diff
@@ -115,15 +115,15 @@ sirf::cReg_NiftiImageDataParameter(const DataHandle* handle, const char* name)
 void*
 sirf::cReg_setRegistrationParameter(void* hp, const char* name, const void* hv)
 {
-    std::shared_ptr<const ImageData> im_sptr;
+    std::shared_ptr<const ImageData<float> > im_sptr;
 
     Registration<float>& s = objectFromHandle<Registration<float> >(hp);
     if (strcmp(name, "reference_image") == 0) {
-        getObjectSptrFromHandle<const ImageData>(hv, im_sptr);
+        getObjectSptrFromHandle<const ImageData<float> >(hv, im_sptr);
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply src/Registration/cReg/cReg_p.cpp:122]: My understanding is that in his C interface Richard has `float` for `dataType`.

File: src/Registration/cReg/NiftyResampler.cpp (hunk 1 above)
Comment (KrisThielemans) [reply src/Registration/cReg/NiftyResampler.cpp]: not sure what you mean. Virtual methods return types have indeed to be either identical or returning a derived version of something related to the base class (I haven't checked exact definition of "covariant return types", see maybe https://stackoverflow.com/questions/25813435/what-exactly-are-covariant-return-types-in-c).

However, that doesn't really apply here as far as I can see. I was merely suggestion to say
```c++
auto it_out = output_to_return_sptr->begin();
```
(if that compiles anyway)

Comment (evgueni-ovtchinnikov) [reply src/Registration/cReg/NiftyResampler.cpp]: AFAIK covariant returns must be either pointers or references

Comment (KrisThielemans) [reply src/Registration/cReg/NiftyResampler.cpp]: agreed. But as it's a reference, it might be possible to just say `auto`. Not sure...

Comment (evgueni-ovtchinnikov) [reply src/Registration/cReg/NiftyResampler.cpp]: from the link you quoted: "Covariance only happens when you are returning a pointer or a reference to a class, and the classes are related by inheritance."

File: src/Registration/cReg/NiftyResampler.cpp [hunk 5]
```diff
@@ -42,18 +42,26 @@ using namespace sirf;
 using namespace detail;
 
 template<class dataType>
-static void convert_ImageData_to_ComplexNiftiImageData(ComplexNiftiImageData<dataType> &output, const std::shared_ptr<const ImageData> input_sptr)
+static void convert_ImageData_to_ComplexNiftiImageData(ComplexNiftiImageData<dataType> &output, const std::shared_ptr<const DataContainer> input_sptr)
```
[the remark this answers is not in the document]
Comment (evgueni-ovtchinnikov) [reply src/Registration/cReg/NiftyResampler.cpp:45]: Here we need a base class for both real and complex `ImageData` - `DataContainer` is the only one we have. Not pretty, but I have not found an alternative.

Regarding shared pointers - it looks like the use of them is not justified in this and likely many other places, but I suggest we create a separate issue on that.

File: src/Registration/cReg/cReg_p.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (KrisThielemans) [reply src/Registration/cReg/cReg_p.cpp:122]: ok. that's what the original code was saying anyway, so I'll resolve this

File: src/xGadgetron/cGadgetron/gadgetron_data_containers.cpp (hunk 3 above)
Comment (evgueni-ovtchinnikov) [reply src/xGadgetron/cGadgetron/gadgetron_data_containers.cpp]: @KrisThielemans works nicely, thanks! Will apply to all container methods after I got rid of all `void*` (operations involving scalars still have them).

File: src/Registration/cReg/NiftyResampler.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (KrisThielemans) [reply src/Registration/cReg/NiftyResampler.cpp:45]: > Here we need a base class for both real and complex `ImageData` - `DataContainer` is the only one we have. Not pretty, but I have not found an alternative.
> 

question is why you need the base class. For instance, as `Resampler` is templated in `dataType`, things like
```
 virtual void set_reference_image(const std::shared_ptr<const DataContainer > reference_image_sptr);

    /// Set floating image. This is the image that would be the floating if you were doing a forward transformation.
    virtual void set_floating_image(const std::shared_ptr<const DataContainer > floating_image_sptr);
```
could all have `shared_ptr<ImageData<dataType>>`. Maybe that helps?

This static function could then easily be templated as well.

> Regarding shared pointers - it looks like the use of them is not justified in this and likely many other places, but I suggest we create a separate issue on that.

sure

[the remark this answers is not in the document]
Comment (KrisThielemans) [reply src/Registration/cReg/NiftyResampler.cpp:45]: in discussions, @evgueni-ovtchinnikov  explained that currently `dataType` is always `float`. Complex images are handled independent. So we need `DataContainer` here at the moment.

**Comment by evgueni-ovtchinnikov** [comment]
Left unattended for a long time, resulting in weird conflicts with master. Taken over by PR #1210.

**Pull request closed without merging**